csharp/src/iceboxnet/ServiceManager.cs

```diff
@@ -441,295 +405,463 @@ async Task StartedAsync(IServiceObserverPrx observer, string[] services)
-                        // configuration file if it's specified with --Ice.Config.
-                        properties.ParseIceArgs(ref info.Args);
+                    // ignore
+                }
 
-                        // Next, parse the service "<service>.*" command line options (the Ice command line options
-                        // were parsed by the createProperties above)
-                        properties.ParseArgs(ref info.Args, service);
-                    }
+                if (info.Communicator is Communicator serviceCommunicator)
+                {
+                    await DestroyServiceCommunicatorAsync(info.Name, serviceCommunicator);
+                }
+            }
 
-                    // Clone the logger to assign a new prefix. If one of the built-in loggers is configured
-                    // don't set any logger.
-                    if (properties.TryGetValue("Ice.LogFile", out string? logFile))
-                    {
-                        logger = _logger.CloneWithPrefix(properties.GetValueOrDefault("Ice.ProgramName") ?? "");
-                    }
+            if (_sharedCommunicator is Communicator sharedCommunicator)
+            {
+                RemoveAdminFacets("IceBox.SharedCommunicator.");
 
-                    // If Admin is enabled on the IceBox communicator, for each service that does not set
-                    // Ice.Admin.Enabled, we set Ice.Admin.Enabled=1 to have this service create facets; then
-                    // we add these facets to the IceBox Admin object as IceBox.Service.<service>.<facet>.
-                    string serviceFacetNamePrefix = "IceBox.Service." + service + ".";
-                    bool addFacets = ConfigureAdmin(properties, serviceFacetNamePrefix);
+                try
+                {
+                    await sharedCommunicator.DisposeAsync();
+                }
+                catch (Exception ex)
+                {
+                    _logger.Warning($"IceBox.ServiceManager: exception while destroying shared communicator:\n{ex}");
+                }
+            }
+            ServicesStopped(stoppedServices);
+        }
 
-                    // Remaining command line options are passed to the communicator. This is necessary for Ice plug-in
-                    // properties (e.g.: IceSSL).
-                    info.Communicator = new Communicator(ref info.Args, properties, logger: logger);
-                    communicator = info.Communicator;
+        private class ServiceInfo
+        {
+            internal Communicator? Communicator { get; } // null means uses shared communicator
 
-                    if (addFacets)
+            /// <summary>Return true when the service is started, and returns false when the service is stopped. Other
+            /// states are not visible to the caller.
+            /// IMPORTANT: This property must be called with ServiceManager's _mutex locked!</summary>
```

Comment:
Did you consider holding a reference to the SreviceManager, we can assign it in ServiceInfo constructor and then just lock the mutex here

Reply:
There are 2 users of this `IsStarted` property:
- StartAsync/StopAsync, which lock the mutex before calling IsStarted on itself
- AddObserverAsync, which locks the mutex before calling IsStarted on all service infos object

For AddObserverAsync, it's critical to get a consistent view, so it's not useful for IsStarted to lock the mutex itself.